source/fe/fe_poly.cc

```diff
@@ -43,13 +43,9 @@ FE_Poly<TensorProductPolynomials<1>, 1, 2>::fill_fe_values(
   dealii::internal::FEValuesImplementation::FiniteElementRelatedData<1, 2>
     &output_data) const
 {
-  // convert data object to internal
-  // data for this class. fails with
-  // an exception if that is not
-  // possible
-  Assert(dynamic_cast<const InternalData *>(&fe_internal) != nullptr,
-         ExcInternalError());
-  const InternalData &fe_data = static_cast<const InternalData &>(fe_internal);
+  // convert data object to internal data for this class. fails with an
+  // exception if that is not possible
+  const InternalData &fe_data = dynamic_cast<const InternalData &>(fe_internal);
```

Comment:
Please leave the assertion in place. It's true that the `dynamic_cast` fails, but with a `std::bad_cast` exception that doesn't reveal anything about the location where the problem happened.

Same for the cases below.

Reply:
I was quite sure that there would some kind of discussion here.

Reply:
:-)

Reply:
As usual for all these PRs, I am happy to incorporate any comments.